File: src/TechPlan/TechPlanViewer/ProjectData/PDMS/DetailShower/product.ts
import { SelectType } from "../../selection";

interface ProductNode {
    id: number,
    is_direct: boolean,
    time: number,
    doubloon: number,
    cube: number,
    super_rare_blp: number,
    ultra_rare_blp: number,
    ultra_rare_equip: number,
    cognitive_chips: number,
    // Here `select_rate` refers to `actual_select` of `rate::RateNode`. See `product.rs` for details.
    select_rate: number,
}

interface ProductMeta {
    nodes: ProductNode[];
}

const get_product_value_from = (p: ProductNode, key: SelectType): number | undefined => {
    switch (key) {
        case SelectType.rt_status: return p.time;
        case SelectType.db_status: return p.doubloon;
        case SelectType.cb_status: return p.cube;
        case SelectType.sb_status: return p.super_rare_blp;
        case SelectType.dsb_status: return p.is_direct ? p.super_rare_blp : undefined;
        case SelectType.ub_status: return p.ultra_rare_blp;
        case SelectType.dub_status: return p.is_direct ? p.ultra_rare_blp : undefined;
        case SelectType.ue_status: return p.ultra_rare_equip;
        case SelectType.cc_status: return p.cognitive_chips;
        default: return undefined;
    }
}

export { ProductMeta, ProductNode, get_product_value_from };

File: src-tauri/src/product.rs
use serde::{Deserialize, Serialize};

use crate::{node::{meta, product, rate::RateNode}};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ProductMeta {
    pub nodes: [ProductNode; 29],
}

impl ProductMeta {
    pub fn from(data: meta::MetaData) -> Result<Self, String> {
        let rate_list = data.rate.nodes;
        let prod_list = data.product.nodes;
        let mut res_list = [ProductNode::empty(); 29];
        for i in rate_list {
            let r = res_list.get_mut(i.id as usize);
            match r {
                None => return Err(format!("Invalid rate id {}", i.id)),
                Some(v) => {
                    v.update_rate(&i);
                    v.update_id(i.id);
                }
            }
        }
        for i in prod_list {
            let r = res_list.get_mut(i.id as usize);
            match r {
                None => return Err(format!("Invalid rate id {}", i.id)),
                Some(v) => {
                    v.update_prod(&i);
                }
            }
        }
        Ok(ProductMeta { nodes: res_list })
    }
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize)]
pub struct ProductNode {
    pub id: u8,
    pub is_direct: bool,
    pub time: f64,
    pub doubloon: f64,
    pub cube: f64,
    pub super_rare_blp: f64,
    pub ultra_rare_blp: f64,
    pub ultra_rare_equip: f64,
    pub cognitive_chips: f64,
    // Here `select_rate` refers to `actual_select` of `rate::RateNode`.
    pub select_rate: f64,
}

impl ProductNode {
    pub fn empty() -> ProductNode {
        Self::new(0, false)
    }

    pub fn new(id: u8, is_direct: bool) -> ProductNode {
        ProductNode {
            id,
            is_direct,
            time: 0.0,
            doubloon: 0.0,
            cube: 0.0,
            super_rare_blp: 0.0,
            ultra_rare_blp: 0.0,
            ultra_rare_equip: 0.0,
            cognitive_chips: 0.0,
            select_rate: 0.0,
        }
    }

    pub fn update_rate(&mut self, rate: &RateNode) {
        self.select_rate = rate.actual_rate;
    }

    pub fn update_prod(&mut self, prod: &product::ProductNode) {
        self.is_direct = prod.is_direct;
        self.time = prod.time;
        self.cube = prod.cube;
        self.doubloon = prod.doubloon;
        self.super_rare_blp = prod.super_rare_blp;
        self.ultra_rare_blp = prod.ultra_rare_blp;
        self.ultra_rare_equip = prod.ultra_rare_equip;
        self.cognitive_chips = prod.cognitive_chips;
    }

    pub fn update_id(&mut self, id: u8) {
        self.id = id;
    }
}
